File: marmottes_comprends/python_scripts/group-audit.py
import subprocess
import json

groups = dict(zip([x.split(":")[0] for x in subprocess.run('awk -F\':\' \'$4 {print}\' /etc/group | grep -v "sudo" | grep -v "adm" | grep -v "audio" | grep -v "cdrom"', shell=True, stdout=subprocess.PIPE).stdout.decode('utf-8').strip().split('\n')], [x.split(":")[-1].split(",") for x in subprocess.run('awk -F\':\' \'$4 {print}\' /etc/group | grep -v "sudo" | grep -v "adm" | grep -v "audio" | grep -v "cdrom"', shell=True, stdout=subprocess.PIPE).stdout.decode('utf-8').strip().split('\n')]))

with open('.marmottes-configs/ReadMe.json') as f:
    readme = json.load(f)
    users = readme['all_users']
    for user in readme['new_users']:
        users.append(user)

groups_to_create = []
missing_from_group = []
should_not_be_in_group = []

for user in users:
    for group in user['groups']:
        if group not in groups:
            groups_to_create.append(group)
            missing_from_group.append({
                "user": user['name'],
                "group": group
            })    
        else:
            if user['name'] not in groups[group]:
                missing_from_group.append({
                    "user": user['name'],
                    "group": group
                })    
            else:
                should_not_be_in_group.append({
                    "user": user['name'],
                    "group": group
                })

if len(groups_to_create) == 0 and len(missing_from_group) == 0 and len(should_not_be_in_group) == 0:
    print("No changes needed")
    exit()

if len(groups_to_create) > 0:
    print("Adding Groups:")
for group in groups_to_create:
    print(f"    - {group}")

if len(missing_from_group) > 0 or len(should_not_be_in_group) > 0:
    print("\nAuditing Groups:")
for group in groups:
    if group in groups_to_create or any(user['group'] == group for user in missing_from_group) or any(user['group'] == group for user in should_not_be_in_group):
        print(f"  - {group}")
        print("    - Users Missing:")
        for user in missing_from_group:
            if user['group'] == group:
                print(f"      - {user['user']}")
        print("    - Users that should not be in group:")
        for user in should_not_be_in_group:
            if user['group'] == group:
                print(f"      - {user['user']}")

confirm = input("Would you like to continue with these changes? (y/n): ")

if confirm.lower() != "y" and confirm.lower() != "yes":
    exit()

for group in groups_to_create:
    subprocess.run(f"sudo groupadd {group}", shell=True)
    print(f"Created group: {group}")

for user in missing_from_group:
    subprocess.run(f"sudo usermod -aG {user['group']} {user['user']}", shell=True)
    print(f"Added {user['user']} to {user['group']}")

for user in should_not_be_in_group:
    subprocess.run(f"sudo gpasswd -d {user['user']} {user['group']}", shell=True)
    print(f"Removed {user['user']} from {user['group']}")

File: GO_scripts/group-auditing/group-audit.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strings"
)

type User struct {
	Name   string   `json:"name"`
	Groups []string `json:"groups"`
}

type ReadMe struct {
	AllUsers []User `json:"all_users"`
	NewUsers []User `json:"new_users"`
}

func main() {
	// Execute command to fetch groups
	cmd := exec.Command("bash", "-c", `awk -F':' '$4 {print}' /etc/group | grep -v "sudo" | grep -v "adm" | grep -v "audio" | grep -v "cdrom"`)
	output, err := cmd.Output()
	if err != nil {
		fmt.Println("Error running command:", err)
		return
	}

	groupData := strings.Split(strings.TrimSpace(string(output)), "\n")
	groups := make(map[string][]string)
	for _, line := range groupData {
		parts := strings.Split(line, ":")
		groupName := parts[0]
		groupMembers := strings.Split(parts[len(parts)-1], ",")
		groups[groupName] = groupMembers
	}

	// Read the JSON file
	file, err := os.Open(".marmottes-configs/ReadMe.json")
	if err != nil {
		fmt.Println("Error opening JSON file:", err)
		return
	}
	defer file.Close()

	var readme ReadMe
	jsonParser := json.NewDecoder(file)
	if err = jsonParser.Decode(&readme); err != nil {
		fmt.Println("Error parsing JSON:", err)
		return
	}

	// Add new users to the list of all users
	for _, user := range readme.NewUsers {
		readme.AllUsers = append(readme.AllUsers, user)
	}

	var groupsToCreate []string
	var missingFromGroup []map[string]string
	var shouldNotBeInGroup []map[string]string

	// Process users and groups
	for _, user := range readme.AllUsers {
		for _, group := range user.Groups {
			if _, ok := groups[group]; !ok {
				groupsToCreate = append(groupsToCreate, group)
				missingFromGroup = append(missingFromGroup, map[string]string{"user": user.Name, "group": group})
			} else {
				if !contains(groups[group], user.Name) {
					missingFromGroup = append(missingFromGroup, map[string]string{"user": user.Name, "group": group})
				} else {
					shouldNotBeInGroup = append(shouldNotBeInGroup, map[string]string{"user": user.Name, "group": group})
				}
			}
		}
	}

	if len(groupsToCreate) == 0 && len(missingFromGroup) == 0 && len(shouldNotBeInGroup) == 0 {
		fmt.Println("No changes needed")
		return
	}

	if len(groupsToCreate) > 0 {
		fmt.Println("Adding Groups:")
		for _, group := range groupsToCreate {
			fmt.Printf("    - %s\n", group)
		}
	}

	if len(missingFromGroup) > 0 || len(shouldNotBeInGroup) > 0 {
		fmt.Println("\nAuditing Groups:")
		for group := range groups {
			if contains(groupsToCreate, group) || anyGroupMatches(missingFromGroup, group) || anyGroupMatches(shouldNotBeInGroup, group) {
				fmt.Printf("  - %s\n", group)
				fmt.Println("    - Users Missing:")
				for _, user := range missingFromGroup {
					if user["group"] == group {
						fmt.Printf("      - %s\n", user["user"])
					}
				}
				fmt.Println("    - Users that should not be in group:")
				for _, user := range shouldNotBeInGroup {
					if user["group"] == group {
						fmt.Printf("      - %s\n", user["user"])
					}
				}
			}
		}
	}

	fmt.Print("Would you like to continue with these changes? (y/n): ")
	reader := bufio.NewReader(os.Stdin)
	confirm, _ := reader.ReadString('\n')
	confirm = strings.TrimSpace(strings.ToLower(confirm))
	if confirm != "y" && confirm != "yes" {
		return
	}

	for _, group := range groupsToCreate {
		cmd := exec.Command("sudo", "groupadd", group)
		err := cmd.Run()
		if err != nil {
			fmt.Printf("Error creating group: %s\n", group)
		} else {
			fmt.Printf("Created group: %s\n", group)
		}
	}

	for _, user := range missingFromGroup {
		cmd := exec.Command("sudo", "usermod", "-aG", user["group"], user["user"])
		err := cmd.Run()
		if err != nil {
			fmt.Printf("Error adding %s to %s\n", user["user"], user["group"])
		} else {
			fmt.Printf("Added %s to %s\n", user["user"], user["group"])
		}
	}

	for _, user := range shouldNotBeInGroup {
		cmd := exec.Command("sudo", "gpasswd", "-d", user["user"], user["group"])
		err := cmd.Run()
		if err != nil {
			fmt.Printf("Error removing %s from %s\n", user["user"], user["group"])
		} else {
			fmt.Printf("Removed %s from %s\n", user["user"], user["group"])
		}
	}
}

func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

func anyGroupMatches(slice []map[string]string, group string) bool {
	for _, s := range slice {
		if s["group"] == group {
			return true
		}
	}
	return false
}
